File: WindowsFormsApplication6/MainForm.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace WindowsFormsApplication6
{
	public partial class MainForm : Form
	{
		public MainForm()
		{
			InitializeComponent();
		}
		byte[] inputBuffer = new byte[1024000];

		private void getHtmlButton_Click(object sender, EventArgs e)
		{
			var ui = TaskScheduler.FromCurrentSynchronizationContext();
			ThreadPool.QueueUserWorkItem(_ => StartRequest(ui));
			getHtmlButton.Enabled = false;
		}

		private void StartRequest(TaskScheduler ui)
		{
			HttpWebRequest webRequest =
				(HttpWebRequest) WebRequest.Create("http://google.ca");
			webRequest.
				BeginGetResponse(
					asyncResult =>
						{
							WebResponse response =
								webRequest.EndGetResponse(asyncResult);
							Stream stream = response.GetResponseStream();
							if (stream == null) return;
							Task<int> bytesRead =
								Task<int>.Factory.
									FromAsync(stream.BeginReadToEnd,
											  stream.EndReadToEnd,
											  inputBuffer, 0, inputBuffer.Length,
											  stream);

							bytesRead.ContinueWith(
								ar =>
									{
										Trace.WriteLine(
											string.Format("Read {0} bytes",
														  ar.Result));
										string text = Encoding.ASCII.
											GetString(inputBuffer, 0,
													  ar.Result);
										SetData(text);
										getHtmlButton.Enabled = true;
										((IDisposable)response).Dispose();
									}, ui);
						}, webRequest);
		}

		private void SetData(string html)
		{
			listBox.Items.Clear();
			if (String.IsNullOrEmpty(html)) return;

			foreach (var x in GetScriptBodies(html))
				listBox.Items.Add(x);
		}

		private static IEnumerable<string> GetScriptBodies(string html)
		{
			const string pattern = @"<script.*?>\s*(?'scriptBody'.+?)\s*</script>";
			MatchCollection matches = Regex.Matches(html, pattern);
			if (matches.Count == 0) yield break;
			foreach (Match match in matches)
			{
				string scriptBody = match.Groups["scriptBody"].Value;
				yield return scriptBody;
			}
		}
	}
}

File: WindowsApplication6/MainForm.vb
Imports System.Net
Imports System.IO
Imports System.Text
Imports System.Text.RegularExpressions
Imports System.Threading.Tasks
Imports System.Threading

Public Class MainForm
    Private ReadOnly finalBuffer(1024000) As Byte

    Private Sub getHtmlButton_Click(sender As System.Object, e As EventArgs) Handles getHtmlButton.Click
        ThreadPool.QueueUserWorkItem(New WaitCallback(AddressOf StartRequest))
        getHtmlButton.Enabled = False
    End Sub

    Private Sub StartRequest()
        Dim inputBuffer(1024000) As Byte
        Dim webRequest As HttpWebRequest = DirectCast(Net.WebRequest.Create("http://google.ca"), HttpWebRequest)
        webRequest.BeginGetResponse(Sub(asyncResult)
                                        Dim response As WebResponse = webRequest.EndGetResponse(asyncResult)
                                        Dim stream As Stream = response.GetResponseStream()
                                        Dim bytesRead As Task(Of Integer) = Task(Of Integer).Factory.FromAsync(
                                            AddressOf stream.BeginReadToEnd, AddressOf stream.EndReadToEnd, inputBuffer, 0, inputBuffer.Length, stream)
                                        Dim ui = TaskScheduler.FromCurrentSynchronizationContext()
                                        bytesRead.ContinueWith(Sub(ar)
                                                                   Trace.WriteLine(String.Format("Read {0} bytes", ar.Result))
                                                                   Dim html = Encoding.ASCII.GetString(inputBuffer, 0, ar.Result)
                                                                   SetData(html)
                                                                   getHtmlButton.Enabled = True
                                                                   DirectCast(response, IDisposable).Dispose()
                                                               End Sub, ui)
                                    End Sub, webRequest)
    End Sub

    Private Sub SetData(ByVal html As String)
        If String.IsNullOrEmpty(html) Then Return

        listBox.Items.Clear()
        For Each x As String In GetScriptBodies(html)
            listBox.Items.Add(x)
        Next
    End Sub

    Private Shared Function GetScriptBodies(ByVal html As String) As IEnumerable(Of String)
        Const pattern As String = "<script.*?>\s*(?'scriptBody'.+?)\s*</script>"
        Dim matches As MatchCollection = Regex.Matches(html, pattern)
        Dim result As List(Of String) = New List(Of String)(matches.Count)
        result.AddRange(From match As Match In matches Select match.Groups("scriptBody").Value)
        Return result
    End Function
End Class
